src/macros.js

```diff
@@ -131,24 +140,85 @@ export default function adMacroReplacement(string, uriEncode, customMacros) {
       videojs.log.warn(`Page variable "${name}" is not a supported type`);
-      return '';
+      pageVariablesMacros[key] = '';
+    } else {
+      pageVariablesMacros[key] = value;
     }
-
-    return uriEncodeIfNeeded(String(value), uriEncode);
   });
 
-  // Replace defaults
-  for (const defaultVal in defaults) {
-    string = string.replace(defaultVal, defaults[defaultVal]);
+  return pageVariablesMacros;
+};
+
+const replaceMacros = function(string, macros, uriEncode, overrides = {}) {
+  for (const macroName in macros) {
+    const resolvedMacroName = overrides.hasOwnProperty(macroName) ? overrides[macroName] : macroName;
+
+    string = string.split(resolvedMacroName).join(uriEncodeIfNeeded(macros[macroName], uriEncode));
+  }
+  return string;
+};
+
+/**
+ *
+ * @param {string} string
+ *                 Any string with macros to be replaced
+ * @param {boolean} uriEncode
+ *                  A Boolean value indicating whether the macros should be replaced with URI-encoded values
+ * @param {object} customMacros
+ *                 An object with custom macros and values to map them to. For example: {'{five}': 5}
+ * @param {boolean} customMacros.disableDefaultMacros
+ *                  A boolean indicating whether replacement of default macros should be forgone in favor of only customMacros
+ * @param {object} customMacros.macroNameOverrides
+ *                 An object that specifies custom names for default macros, following the following format:
+ *                 // {'{default-macro-name}': '{new-macro-name}'}
+ *                 {'{player.id}': '{{PLAYER_ID}}', ...}
+ * @returns {string}
+ *          The provided string with all macros replaced. For example: adMacroReplacement('{player.id}') returns a string of the player id
+ */
+export default function adMacroReplacement(string, uriEncode = false, customMacros = {}) {
+  const disableDefaultMacros = customMacros.disableDefaultMacros || false;
+  const macroNameOverrides = customMacros.macroNameOverrides || {};
+
+  // Remove special properties from customMacros
+  delete customMacros.disableDefaultMacros;
+  delete customMacros.macroNameOverrides;
+
+  const macros = customMacros;
+  const {string: updatedString, macrosToOverride} = processMacroNameOverrides(string, macroNameOverrides);
+
+  string = updatedString;
+
+  if (disableDefaultMacros) {
+    return replaceMacros(string, macros, uriEncode, macrosToOverride);
+  }
+
+  // Get macros with defaults e.g. {x=y}, store the values in `defaults` and replace with standard macros in the string
+  const {defaults, modifiedString} = getDefaultValues(string);
+
+  string = modifiedString;
+
+  // Get all macro values
+  Object.assign(macros, getStaticMacros.call(this));
+  Object.assign(macros, getMediaInfoMacros(this.mediainfo, defaults));
+  Object.assign(macros, getTcfMacros(tcData));
+  Object.assign(macros, getPageVariableMacros(string, defaults));
```

Comment:
nitpick/non-blocking: These could be a single `Object.assign()` call as it can take any number of arguments.

Reply:
Lol yup good point